wandb/sdk/launch/runner/gcp_vertex.py

```diff
@@ -0,0 +1,244 @@
+import datetime
+import os
+import subprocess
+import time
+from typing import Any, Dict, List, Optional
+
+from google.cloud import aiplatform  # type: ignore
+from six.moves import shlex_quote
+import wandb
+from wandb.errors import CommError, LaunchError
+import yaml
+
+from .abstract import AbstractRun, AbstractRunner, Status
+from .._project_spec import (
+    get_entry_point_command,
+    LaunchProject,
+)
+from ..docker import (
+    build_docker_image_if_needed,
+    construct_gcp_image_uri,
+    docker_image_exists,
+    docker_image_inspect,
+    generate_docker_base_image,
+    pull_docker_image,
+    validate_docker_installation,
+)
+from ..utils import (
+    PROJECT_DOCKER_ARGS,
+    PROJECT_SYNCHRONOUS,
+)
+
+GCP_CONSOLE_URI = "https://console.cloud.google.com"
+
+
+def docker_push(image: str) -> None:
+    # tmp function until sagemaker pr merged
+    subprocess.run(["docker", "push", image])
+
+
+class VertexSubmittedRun(AbstractRun):
+    def __init__(self, job: Any) -> None:
+        self._job = job
+
+    @property
+    def id(self) -> str:
+        # numeric ID of the custom training job
+        return self._job.name  # type: ignore
+
+    @property
+    def name(self) -> str:
+        return self._job.display_name  # type: ignore
+
+    @property
+    def gcp_region(self) -> str:
+        return self._job.location  # type: ignore
+
+    @property
+    def gcp_project(self) -> str:
+        return self._job.project  # type: ignore
+
+    def get_page_link(self) -> str:
+        return "{console_uri}/vertex-ai/locations/{region}/training/{job_id}?project={project}".format(
+            console_uri=GCP_CONSOLE_URI,
+            region=self.gcp_region,
+            job_id=self.id,
+            project=self.gcp_project,
+        )
+
+    def wait(self) -> bool:
+        self._job.wait()
+        return self.get_status().state == "finished"
+
+    def get_status(self) -> Status:
+        job_state = str(self._job.state)  # extract from type PipelineState
+        if job_state == "PipelineState.PIPELINE_STATE_SUCCEEDED":
+            return Status("finished")
+        if job_state == "PipelineState.PIPELINE_STATE_FAILED":
+            return Status("failed")
+        if job_state == "PipelineState.PIPELINE_STATE_RUNNING":
+            return Status("running")
+        return Status("unknown")
+
+    def cancel(self) -> None:
+        self._job.cancel()
+
+
+class VertexRunner(AbstractRunner):
+    """Runner class, uses a project to create a VertexSubmittedRun"""
+
+    def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
+        resource_args = launch_project.resource_args
+        gcp_config = get_gcp_config(resource_args.get("gcp_config") or "default")
+        gcp_project = (
+            resource_args.get("gcp_project")
+            or gcp_config["properties"]["core"]["project"]
+        )
+        gcp_zone = resource_args.get("gcp_region") or gcp_config["properties"].get(
```

Comment:
Should we have the users specify aws_region, and gcp_region or just region?

Reply:
nonissue now that we have resource args separated by resource key